Sources/NIOEmbedded/AsyncTestingChannel.swift

```diff
@@ -0,0 +1,508 @@
+    ///
+    /// The `NIOAsyncTestingChannel` will store an error if some error travels the `ChannelPipeline` all the way past its end.
+    public func throwIfErrorCaught() async throws {
+        try await self.embeddedEventLoop.executeInContext {
+            try self._throwIfErrorCaught()
+        }
+    }
+
+    @usableFromInline
+    func _throwIfErrorCaught() throws {
+        self.embeddedEventLoop.preconditionInEventLoop()
+        if let error = self.channelcore.error {
+            self.channelcore.error = nil
+            throw error
+        }
+    }
+
+    @inlinable
+    func _readFromBuffer<T>(buffer: inout CircularBuffer<NIOAny>) throws -> T? {
+        self.embeddedEventLoop.preconditionInEventLoop()
+
+        if buffer.isEmpty {
+            return nil
+        }
+        let elem = buffer.removeFirst()
+        guard let t = self._channelCore.tryUnwrapData(elem, as: T.self) else {
+            throw WrongTypeError(expected: T.self, actual: type(of: self._channelCore.tryUnwrapData(elem, as: Any.self)!))
+        }
+        return t
+    }
+
+    /// Create a new instance.
+    ///
+    /// During creation it will automatically also register itself on the `NIOAsyncTestingEventLoop`.
+    ///
+    /// - parameters:
+    ///     - loop: The `NIOAsyncTestingEventLoop` to use.
+    public init(loop: NIOAsyncTestingEventLoop = NIOAsyncTestingEventLoop()) {
+        self.embeddedEventLoop = loop
+        self._pipeline = ChannelPipeline(channel: self)
+        self.channelcore = EmbeddedChannelCore(pipeline: self._pipeline, eventLoop: self.eventLoop)
+    }
+
+    /// Create a new instance.
+    ///
+    /// During creation it will automatically also register itself on the `NIOAsyncTestingEventLoop`.
+    ///
+    /// - parameters:
+    ///     - handler: The `ChannelHandler` to add to the `ChannelPipeline` before register.
+    ///     - loop: The `NIOAsyncTestingEventLoop` to use.
+    public convenience init(handler: ChannelHandler, loop: NIOAsyncTestingEventLoop = NIOAsyncTestingEventLoop()) async {
+        await self.init(handlers: [handler], loop: loop)
+    }
+
+    /// Create a new instance.
+    ///
+    /// During creation it will automatically also register itself on the `NIOAsyncTestingEventLoop`.
+    ///
+    /// - parameters:
+    ///     - handlers: The `ChannelHandler`s to add to the `ChannelPipeline` before register.
+    ///     - loop: The `NIOAsyncTestingEventLoop` to use.
+    public convenience init(handlers: [ChannelHandler], loop: NIOAsyncTestingEventLoop = NIOAsyncTestingEventLoop()) async {
+        self.init(loop: loop)
+
+        try! await self._pipeline.addHandlers(handlers)
+
+        // This will never throw...
+        try! await self.register()
+    }
+
+    /// - see: `Channel.setOption`
+    @inlinable
+    public func setOption<Option: ChannelOption>(_ option: Option, value: Option.Value) -> EventLoopFuture<Void> {
+        if self.eventLoop.inEventLoop {
```

Comment:
Should we just copy the implementation of `setOptionSync` and `fatalError` here?

Reply:
I'm inclined to leave this as is: it's a common pattern.